rust/src/coordinator.rs

```diff
@@ -105,1045 +119,590 @@ impl RoundSeed {
 
     /// Update the seed round parameter.
     fn update_round_seed(&mut self) {
         // safe unwrap: `sk` and `seed` have same number of bytes
-        let (_, sk) =
-            SigningKeySeed::from_slice_unchecked(self.sk().as_slice()).derive_signing_key_pair();
+        let (_, sk) = SigningKeySeed::from_slice_unchecked(self.state.sk.as_slice())
+            .derive_signing_key_pair();
         let signature = sk.sign_detached(
             &[
-                self.seed().as_slice(),
-                &self.sum().to_le_bytes(),
-                &self.update().to_le_bytes(),
+                self.state.seed.as_slice(),
+                &self.state.sum.to_le_bytes(),
+                &self.state.update.to_le_bytes(),
             ]
             .concat(),
         );
         // safe unwrap: length of slice is guaranteed by constants
-        *self.seed_mut() =
+        self.state.seed =
             RoundSeed::from_slice_unchecked(sha256::hash(signature.as_slice()).as_ref());
     }
 
-    /// Check whether enough sum participants submitted their ephemeral keys to start the update
-    /// phase.
-    fn has_enough_sums(&self) -> bool {
-        self.sum_dict().len() >= *self.min_sum()
-    }
-
-    /// Check whether enough update participants submitted their models and seeds to start the sum2
-    /// phase.
-    fn has_enough_seeds(&self) -> bool {
-        self.seed_dict()
-            .values()
-            .next()
-            .map(|dict| dict.len() >= *self.min_update())
-            .unwrap_or(false)
+    /// Prepare the coordinator for a new round and go back to the initial phase.
+    async fn start_new_round(&mut self) {
+        self.clear_redis_state().await; // remove only the dicts
+        self.update_round_thresholds();
+        self.update_round_seed();
+        self.state.phase = Phase::Idle;
+        self.set_state().await;
+        self.sum_phase_cache = None;
     }
 
-    /// Check whether enough sum participants submitted their masks to start the idle phase.
-    fn has_enough_masks(&self) -> bool {
-        let mask_count = self.mask_dict().values().sum::<usize>();
-        mask_count >= *self.min_sum()
+    /// End the sum2 phase and proceed to the idle phase to end the round.
+    async fn proceed_idle_phase(&mut self) {
+        info!("going to idle phase");
+        self.start_new_round().await;
     }
 
     /// End the idle phase and proceed to the sum phase to start the round.
-    fn proceed_sum_phase(&mut self) {
+    async fn proceed_sum_phase(&mut self) {
         info!("going to sum phase");
         self.gen_round_keypair();
-        *self.phase_mut() = Phase::Sum;
-        self.emit_event(ProtocolEvent::StartSum(self.round_parameters()));
+        self.state.phase = Phase::Sum;
+        self.set_state().await;
+
+        // We want to process as many messages as possible at the same time but we don't want to
+        // process more than is necessary (min_sum).
```

Comment:
we need to be careful about two things here:
1) if the coordinator only takes the bare minimum of `sum` participants in the `Sum` phase (meaning around `min_sum` many) and then too many of those drop out and are unavailable during the `Sum2` phase, then the round will abort. therefore we should choose a safety margin, maybe a certain percentage, or split the parameters into `min_sum` and `min_sum2` where `min_sum > min_sum2`.
2) if the coordinator selects only a subset of available participants, then selection bias must be avoided. for example only picking the first `m` out of `n` available participants or having a too short phase. not only does this mean a selection bias for faster devices/hardware but it might also lead to attacks where eligible participants gain an advantage by having better latency or processing power. a way to do this might be to have a reasonably large time frame for the phases (i know that this is a vague term, it should be configurable in the end) and to select a random subset of all available participants for that phase if there are to many of them.

if we play strictly by the rules of the protocol we wouldn't do any of the above but just abort the round and adjust the `sum` and `update` fractions for the next round accordingly. this avoids the issues explained above at the potential cost of some overhead, but in the end selecting good initial parameters for the protocol is the responsibility of the people setting up a specific federated learning use case.

Reply:
Thanks for the comment.

1. If I understand it correctly, the coordinator can still unmask the global model, even if not all sum participants (who responded in the sum phase) do not respond in the sum2 phase?

Reply:
exactly. technically a single valid `sum2` message suffices to unmask the global model. all sum participants are doing just redundant work. the main reasons for this are resiliency against sum participant dropouts and against malicious sum participants.